test/unit/configuration-test.js

```diff
@@ -156,9 +162,409 @@ describe('configuration.applyLoggingOptions()', () => {
     it('the application logger level is set to debug', () => {
       assert.equal(logger.transports.console.level, 'debug');
     });
-
     it('the application logger is set to add timestamps', () => {
       assert.isTrue(logger.transports.console.timestamp);
     });
   });
 });
+
+
+describe('configuration._coerceRemovedOptions()', () => {
```

Comment:
I don't like the way we are testing the specific unit for a few reasons (firstly we have to export this internal method for testing), but the biggest reason is it is perhaps error prone. Especially considering Dredd does not error out for invalid options. For example, if I invoke dredd with `--foo` or even a typo of a valid option, dredd will continue instead of erroring out:

```
$ dredd api-description.apib http://127.0.0.1:3000 --dry-run --foo
info: Beginning Dredd testing...
...
complete: Tests took 5ms
```

IMO this is a huge anti pattern because if a user typos or misenters an option it could result in undesired behaviour which could potentially be destructive.

I'll show you perhaps an example outside of Dredd scope:

```shell
$ rm -fr --dry-run /
rm: illegal option -- -
usage: rm [-f | -i] [-dPRrvW] file ...
       unlink file
```

Now, if `--dry-run` (perhaps I typoed `--dri-run`) doesn't exist the potentially destructive command did not proceed. An example closer to Dredd could be perhaps attempting to use Dredd CLI options to skip certain destructive methods. For example if I wanted to run Dredd on production and wanted to skip method DELETE for obvious reasons and I typoed the skip command I would hope that Dredd errors out due to invalid arguments instead of running a mis-understood command.

---

Anyway, my actual point (sorry got distracted on side-bug story/tangent 😀) is what if `_coerceRemovedOptions` created a invalid option that doesn't exist we would never know. Since the test is for the unintegrated unit (and also Dredd doesn't validate invalid options) we could have room for invalid code. Perhaps right now we can be 100% sure with careful review, but what about in a few years times when (future) you or someone else decides to remove an option `--color` because it is spelt wrong :trollface: and they add `--colour` these tests would continue to work so they cannot protect against this type of bug.

Reply:
Anyway, tl;dr I am not sure these specific tests add much value over what the code does. They do not protect you against future refactoring.

Reply:
I have to admit I've written the tests _ex post_. They were useful in catching bugs in the function I wrote prior to them 😓 

Your points are completely valid. I consider the whole file to be a brownfield. The fact Dredd is in general silent about mistyped options is terrible, but it is an existing behavior. It's solvable probably only by switching the CLI options parser and changing how the configuration is processed 😞

Perhaps `_coerceRemovedOptions` could have been checked also by an integration test, but ideally I remove all the checks after a period of time with a major release and introduce a better configuration management for Dredd (#1101).

Reply:
tl;dr I completely agree, but I think fixing this is out of the scope of this PR. I added ugly things, I know it, and I hope I'll have time to improve the CLI later. And I hope if I won't have time to do it, future generations of people working on Dredd at least find these comments and won't hate me that much 🙏 😄